### PR TITLE
Update the Poetry Core version in the Habushu project to the required version

### PR DESCRIPTION
**Pull request by satishmanda**
AIOPS-2767 Added functionality to upgrade the poetry-core version required by the Habushu

### CODE REVIEW

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            tomlFileConfig.load();
+            Optional<Config> toolBuildSystem = tomlFileConfig.getOptional(TomlUtils.BUILD_SYSTEM);
+            if (toolBuildSystem.isPresent()) {
+                Config buildSystem = toolBuildSystem.get();
+                Map<String, Object> dependencyMap = buildSystem.valueMap();
+                for (Map.Entry<String, Object> dependency : dependencyMap.entrySet()) {
+                    isMajorVersionUpdateRequired = false;
+                    isMinorVersionUpdateRequired = false;
+                    // check if we need to upgrade the poetry-core version.
+                   if(IsPoetrycoreUpgradeRequired(dependency)) {
```
Comment (cpointe-ibllanos) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: A: Minor, the if statement is unnecessary.
`return IsPoetrycoreUpgradeRequired(dependency)`

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                            }
+
+                            // update the poetry-core minor version if required.
+                            if (isMinorVersionUpdateRequired) {
+                                int minorVersionIndex = line.indexOf(TomlUtils.DOT) + 1;
+                                int endMinorVersionIndex = line.lastIndexOf(TomlUtils.DOT);
+                                String strPoetryCoreReqMinorVersion = PoetryUtil.getMinorReqVersion(PoetryUtil.POETRY_CORE_VERSION_REQUIREMENT);
+                                stringBuilder = stringBuilder.replace(minorVersionIndex, endMinorVersionIndex, strPoetryCoreReqMinorVersion);
+                                line = stringBuilder.toString();
+                                logger.info("Updating the Poetry-core major version to {}.", strPoetryCoreReqMinorVersion);
```
Comment (cpointe-ibllanos) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: A: Typo? The string refers to the major version even though this is for the minor. 

Also, would the logs on line 99 and 89 be better suited for logger.debug? Doesn't make sense to say it's upgrading twice in the info logs every time; I think the single info statement on line 101 is the most helpful info log for a normal build.

**Review by cpointe-ibllanos — APPROVED** [review]
Looks good to me, just minor comments. Thanks for utilizing good practices like clear variable naming, properly defined constants, and clear test steps.

I'd get a second opinion since I'm pretty new to Habushu's codebase however.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+
+/**
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java:25]: A (Administrative): minor, but best if we don't note the versions here as they will likely drift when we upgrade the constants but miss them here.  Recommend you refine this to say "inline with major and minor versions specified in `PoetryUtil`" or similar.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java (hunk 2 above)
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: S (Substantive): comment says major, but seems like it should be minor.

Comment (d-ryan-ashcraft) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: Seems like 3 info level output messages is a little heavy too - maybe the major and minor lines should become debug or be removed?

Also seems like major should spit out the version it is updating to for consistency, if kept.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                            logger.info("Updated the Poetry-core version to {} required by Habushu.", PoetryUtil.POETRY_CORE_VERSION_REQUIREMENT);
+                        }
+                    }
+                }
+                fileContent += line + "\n";
+                line = reader.readLine();
+            }
+
+        } catch (IOException e) {
+            throw new HabushuException("Problem reading pyproject.toml while updating version for poetry-core dependencies!", e);
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: S: This and the next exception - seems like `poetry-core dependencies` should be something like `the build system's poetry-core version`

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+        try {
+            TomlUtils.writeTomlFile(pyProjectTomlFile, fileContent);
+        } catch (IOException e) {
+            throw new BatonException("Problem while writing dependencies to TomlFile during upgrade of poetry-core version!", e);
+        }
+        return true;
+    }
+
+    private boolean IsPoetrycoreUpgradeRequired(Map.Entry<String, Object> dependency) {
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: S: `IsPoetrycoreUpgradeRequired" should be lowercased to "isPoetrycoreUpgradeRequired"

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+                if ((poetryCoreMajorVer == poetryCoreReqMajorVersion && poetryCoreMinorVer < poetryCoreReqMinorVersion)) {
+                    isMinorVersionUpdateRequired = true;
+                }
+
+                // Return true if major or minor version upgarde is required.
+                if(isMajorVersionUpdateRequired || isMinorVersionUpdateRequired) {
+                    TomlReplacementTuple replacementTuple = new TomlReplacementTuple(packageName, poetrycoreVer, "poetry-core>=1.6.0");
+                    replacements.put(packageName, replacementTuple);
+                    logger.info("Found Poetry-core dependency with version : {} which is less than the required version of 1.6.0 or higher for Habushu. Poetry-core version dependency will be updated to 1.6.0.", poetrycoreVer);
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: A: seems like a 4th place this is getting logged - let's keep it tidy, 1 info level log.  We can be more verbose at debug level, if it makes sense.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/util/PoetryUtil.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public static final String POETRY_VERSION_REQUIREMENT = "^1.5.0";
+
+    /**
+     * Specifies the semver compliant requirement for the version of Poetry-core that
+     * must be installed and available for Habushu to use.
+     */
+    public static final String POETRY_CORE_VERSION_REQUIREMENT = "^1.6.0";
+
+    /**
```
Comment (d-ryan-ashcraft) [inline habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/util/PoetryUtil.java]: S: these don't seem like Poetry related Util methods, maybe they belong elsewhere?

**Review by d-ryan-ashcraft — CHANGES_REQUESTED** [review]

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java (hunk 6 above)
Comment (satishmanda) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: Yes. It should be at the debug level. Made the change to log as debug.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/util/PoetryUtil.java (hunk 7 above)
Comment (satishmanda) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/util/PoetryUtil.java]: I am thinking of adding them to the TomlUtil class as these helper methods can be used to extract major and minor version.

File: habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java (hunk 1 above)
Comment (satishmanda) [reply habushu-maven-plugin/src/main/java/org/technologybrewery/habushu/migration/CustomPoetrycoreVersionMigration.java]: We are looping through the dependency map and iterate through all the elements to see if we any of it requires the upgrade. We need the if statement here incase any one of them requires upgrade.

**Review by d-ryan-ashcraft — APPROVED** [review]

**Pull request merged**